solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1021,8 +1047,8 @@ public DocSet getAndCacheDocSet(Query query, SolrCache<Query, DocSet> cache) thr
     }
 
     DocSet answer;
-    QueryLimits queryLimits = QueryLimits.getCurrentLimits();
-    if (queryLimits.isLimitsEnabled()) {
+    if (!OPTIMISTIC_QUERY_LIMITED_CACHE_COMPUTATION
+        && QueryLimits.getCurrentLimits().isLimitsEnabled()) {
```

Comment:
Because under the hood our current cache implementation actually still uses get-then-put, these 2 lines are actually the only practically relevant part of the change atm. All of the concerns called out below apply only to `computeIfAbsent()` implementations that in practice actually block on single/shared computation among different threads.

As such the main thing that this does is pass a lambda that gives the cache implementation some extra control wrt how entries are constructed, which is all we need for seg-aware caching to work.